### ISSUE TITLE
MicroProfile Config Load Error

### ISSUE THREAD

**Issue by hantsy** [eclipse-ee4j/glassfish#25298]
Windows 10, Java 21 , Glassfish 8.0.0-M9

Created `@ConfigProperties` class, and got exception when deploying phase.

```bash
1. com.sun.enterprise.module.ResolveError: Failed to start OSGiModuleImpl:: Bundle = [org.glassfish.main.microprofile.config [91]], State = [NEW]Unable to resolve
    org.glassfish.main.microprofile.config [91]
    missing requirement
[91] 
jar = file:/home/runner/work/jakartaee11-sandbox/jakartaee11-sandbox/cdi/target/glassfish8/glassfish/modules/microprofile-config.jar
artifactId = microprofile-config
groupId = org.glassfish.main.microprofile
version = 8.0.0-M9


2. java.lang.IllegalStateException: Could not load descriptor SystemDescriptor(
	implementation=org.glassfish.microprofile.config.ConfigContainer
	name=org.glassfish.microprofile.config.ConfigContainer
	contracts={org.glassfish.microprofile.config.ConfigContainer,org.glassfish.api.container.Container}
	scope=jakarta.inject.Singleton
	qualifiers={}
	descriptorType=CLASS
	descriptorVisibility=NORMAL
	metadata=Bundle-SymbolicName={org.glassfish.main.microprofile.config},Bundle-Version={8.0.0.M9}
	rank=0
	loader=OsgiPopulatorPostProcessor.HK2Loader(OSGiModuleImpl:: Bundle = [org.glassfish.main.microprofile.config [91]], State = [NEW],98086928)
	proxiable=null
	proxyForSameScope=null
	analysisName=null
	id=956
	locatorId=0
	identityHashCode=174473294
	reified=false)
3. java.lang.IllegalStateException: Unable to resolve
    org.glassfish.main.microprofile.config [91]
    missing requirement
        package = io.helidon.common.types
```

**Comment by hantsy** [comment]
Add `io.helidon.microprofile.config:helidon-microprofile-config` and `io.helidon.common:helidon-common-types` in the project dependencies, still report the issue.

```xml
<dependency>
    <groupId>io.helidon.microprofile.config</groupId>
    <artifactId>helidon-microprofile-config</artifactId>
    <version>4.1.6</version>
</dependency>
<!-- https://mvnrepository.com/artifact/io.helidon.common/helidon-common-types -->
<dependency>
    <groupId>io.helidon.common</groupId>
    <artifactId>helidon-common-types</artifactId>
    <version>4.1.6</version>
</dependency>
```

**Comment by renatsaf** [comment]
Hi @OndroMih I. I asked Claude to analyze the issue.

**Comment by renatsaf** [comment]
This is an OSGi bundle-assembly problem in the `microprofile-config` module (`appserver/microprofile/config/pom.xml`), not an application issue — which is why adding `helidon-microprofile-config` / `helidon-common-types` to the *application* dependencies has no effect. The unresolved bundle lives in the GlassFish module/OSGi layer and is resolved independently of the app classloader.

**Root cause.** `microprofile-config.jar` is built as a fat OSGi bundle by the Felix `maven-bundle-plugin` (bnd), embedding the Helidon MP Config tree via `Embed-Dependency` + `Embed-Transitive`. The `Import-Package` instruction ends in `*`, so bnd generates an import for every package the embedded code references — including `io.helidon.common.types`. Nothing in `glassfish/modules/` exports that package, so the bundle can't be resolved (`State = [NEW]`), which then cascades into the `Could not load descriptor ... ConfigContainer` and `missing requirement package = io.helidon.common.types` errors.

Note that `helidon-common-types` was already listed in `Embed-Dependency` at M9, so this wasn't simply a forgotten entry — the embedded Helidon tree was version-drifting and being assembled inconsistently across Helidon 4.x package reshuffles:

| Build | `helidon-config.version` | helidon-bom imported | Result |
|-------|--------------------------|----------------------|--------|
| 8.0.0-M9 | 4.1.3 | no (unmanaged versions) | fails |
| 8.0.0-M12 | 4.2.2 | yes | still fails |
| current (8.0.3-SNAPSHOT) | 4.4.1 | yes | resolves |

**Fix history** (iterative):
- `216004fb` (2025-01-15) "Update Helidon config to actually work again"
- `40c6cbbe` (2025-07-24) "Fix microprofile config again" — added the `io.helidon:helidon` aggregator to the embed list
- `205dbb90` (2026-01-10) "Helidon 4.3.3 and fix" — dropped `helidon-inject-api`/`helidon-common-service-loader`, added `helidon-service-registry`/`helidon-metadata`
- `ad444eb7` (2026-03-21) "Add new Helidon dependencies into the OSGi bundle" — added `helidon-common-crypto`, `helidon-common-key-util`, `helidon-config-encryption`

These landed in GA **8.0.0** (2026-02-04) and were refined through **8.0.1** / **8.0.2**. The M-series milestone builds (M9–M12) are affected; current 8.0.x (Helidon 4.4.1) resolves the bundle.

This issue is still open — if maintainers can confirm against an 8.0.x GA build, it looks ready to close.

**Comment by OndroMih** [comment]
Yes, I believe this is already fixed in GF 8.0.0, it was reported only for milestone versions.

**Comment by renatsaf** [comment]
Should I test @hantsy 's app on the last GF?

**Comment by renatsaf** [comment]
Verified on **GlassFish 8.0.2** (Helidon 4.4.1) — the fix holds.

I ran @hantsy's `cdi` module with the previously `@Disabled` `ApplicationConfigTest` re-enabled, deploying the `@ConfigProperties`-annotated `ApplicationProperties` WAR via the Arquillian managed container:

```
Tests run: 1, Failures: 0, Errors: 0, Skipped: 0 -- in com.example.it.ApplicationConfigTest
Get application info result: Hello Hantsy from test name
```

The app deployed and the `@ConfigProperties` injection worked. The GlassFish `server.log` contains **no** `ResolveError`, `microprofile-config` bundle failure, or `missing requirement package = io.helidon.common.types` — the `microprofile-config` OSGi bundle resolves cleanly.

So this is resolved on the 8.0.x GA line; it only affected the M-series milestone builds (M9–M12).

**Comment by renatsaf** [comment]
@OndroMih the task can be closed. See the previous comment.
